{
  "comments": [
    {
      "key": {
        "uuid": "8f2e4698_cf11b2df",
        "filename": "src/main/java/com/googlesource/gerrit/plugins/multisite/validation/BatchRefUpdateValidator.java",
        "patchSetId": 9
      },
      "lineNbr": 105,
      "author": {
        "id": 1006192
      },
      "writtenOn": "2020-01-08T21:15:58Z",
      "side": 1,
      "message": "This is the *real* split-brain and we should not return a LOCK_FAILURE in that case. Correct?\nRetrying wouldn\u0027t help.",
      "range": {
        "startLine": 105,
        "startChar": 34,
        "endLine": 105,
        "endChar": 63
      },
      "revId": "38b5635d24d1809085d64ffe247e76440d0db296",
      "serverId": "173816e5-2b9a-37c3-8a2e-48639d4f1153",
      "unresolved": true
    },
    {
      "key": {
        "uuid": "12fb66f5_508d0245",
        "filename": "src/main/java/com/googlesource/gerrit/plugins/multisite/validation/BatchRefUpdateValidator.java",
        "patchSetId": 9
      },
      "lineNbr": 105,
      "author": {
        "id": 1054778
      },
      "writtenOn": "2020-01-09T10:09:19Z",
      "side": 1,
      "message": "I think both exceptions might highlight a real split-brain:\n- OutOfSyncException is thrown when, doing a ref-update, we find out that we are not in sync with the shared-ref. It is definitively worth retry since the node might just be behind a bit.\n- SharedDbSplitBrainException is a generic RuntimeException thrown by the Shared RefDB implementation. It could be worth retrying, it really depends on why the exception was raised. I think though that maybe the retry mechanism in this case shouldn\u0027t be done by the client, but by the multi-site plugin, hence I am going to remove the catching of the exception from here.\n\nAs a side note I think we should review the alert we have on the metrics we collect. At the moment we only have alerts generated from \"OutOfSynceException\"",
      "parentUuid": "8f2e4698_cf11b2df",
      "range": {
        "startLine": 105,
        "startChar": 34,
        "endLine": 105,
        "endChar": 63
      },
      "revId": "38b5635d24d1809085d64ffe247e76440d0db296",
      "serverId": "173816e5-2b9a-37c3-8a2e-48639d4f1153",
      "unresolved": false
    },
    {
      "key": {
        "uuid": "3daa217f_63577182",
        "filename": "src/main/java/com/googlesource/gerrit/plugins/multisite/validation/BatchRefUpdateValidator.java",
        "patchSetId": 9
      },
      "lineNbr": 109,
      "author": {
        "id": 1006192
      },
      "writtenOn": "2020-01-08T21:15:58Z",
      "side": 1,
      "message": "out of sync with the shared ref-db.",
      "range": {
        "startLine": 109,
        "startChar": 56,
        "endLine": 109,
        "endChar": 67
      },
      "revId": "38b5635d24d1809085d64ffe247e76440d0db296",
      "serverId": "173816e5-2b9a-37c3-8a2e-48639d4f1153",
      "unresolved": true
    },
    {
      "key": {
        "uuid": "5e033416_be92654c",
        "filename": "src/main/java/com/googlesource/gerrit/plugins/multisite/validation/BatchRefUpdateValidator.java",
        "patchSetId": 9
      },
      "lineNbr": 109,
      "author": {
        "id": 1054778
      },
      "writtenOn": "2020-01-09T10:09:19Z",
      "side": 1,
      "message": "Done",
      "parentUuid": "3daa217f_63577182",
      "range": {
        "startLine": 109,
        "startChar": 56,
        "endLine": 109,
        "endChar": 67
      },
      "revId": "38b5635d24d1809085d64ffe247e76440d0db296",
      "serverId": "173816e5-2b9a-37c3-8a2e-48639d4f1153",
      "unresolved": false
    },
    {
      "key": {
        "uuid": "7bc4a124_37c66611",
        "filename": "src/main/java/com/googlesource/gerrit/plugins/multisite/validation/RefUpdateValidator.java",
        "patchSetId": 9
      },
      "lineNbr": 130,
      "author": {
        "id": 1006192
      },
      "writtenOn": "2020-01-08T21:15:58Z",
      "side": 1,
      "message": "This is the *real* split-brain situation, correct? Retrying wouldn\u0027t help and we should not return LOCK_FAILURE IMHO.",
      "range": {
        "startLine": 130,
        "startChar": 32,
        "endLine": 130,
        "endChar": 63
      },
      "revId": "38b5635d24d1809085d64ffe247e76440d0db296",
      "serverId": "173816e5-2b9a-37c3-8a2e-48639d4f1153",
      "unresolved": true
    },
    {
      "key": {
        "uuid": "a89dfc7d_aad042f7",
        "filename": "src/main/java/com/googlesource/gerrit/plugins/multisite/validation/RefUpdateValidator.java",
        "patchSetId": 9
      },
      "lineNbr": 130,
      "author": {
        "id": 1083454
      },
      "writtenOn": "2020-01-09T08:15:45Z",
      "side": 1,
      "message": "Ok so my understanding of this feature was a bit different. I thought that we want to differentiate between server error(500) and split brain(LOCK_FAILURE) to improve user experience. And then have retry mechanism which will retry multiple times with backoff and give up if no result. So for OutOfSync it will eventually succeed but SharedDbSplit not.",
      "parentUuid": "7bc4a124_37c66611",
      "range": {
        "startLine": 130,
        "startChar": 32,
        "endLine": 130,
        "endChar": 63
      },
      "revId": "38b5635d24d1809085d64ffe247e76440d0db296",
      "serverId": "173816e5-2b9a-37c3-8a2e-48639d4f1153",
      "unresolved": true
    },
    {
      "key": {
        "uuid": "cd508e6c_419bc47b",
        "filename": "src/main/java/com/googlesource/gerrit/plugins/multisite/validation/RefUpdateValidator.java",
        "patchSetId": 9
      },
      "lineNbr": 130,
      "author": {
        "id": 1054778
      },
      "writtenOn": "2020-01-09T10:09:19Z",
      "side": 1,
      "message": "Done",
      "parentUuid": "a89dfc7d_aad042f7",
      "range": {
        "startLine": 130,
        "startChar": 32,
        "endLine": 130,
        "endChar": 63
      },
      "revId": "38b5635d24d1809085d64ffe247e76440d0db296",
      "serverId": "173816e5-2b9a-37c3-8a2e-48639d4f1153",
      "unresolved": false
    }
  ]
}